dbms/src/Storages/DeltaMerge/VersionChain/RowKeyFilter.cpp

```diff
@@ -0,0 +1,310 @@
+// Copyright 2025 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Interpreters/Context.h>
+#include <Storages/DeltaMerge/ColumnFile/ColumnFileDataProvider.h>
+#include <Storages/DeltaMerge/DMContext.h>
+#include <Storages/DeltaMerge/File/DMFileBlockInputStream.h>
+#include <Storages/DeltaMerge/File/DMFilePackFilter.h>
+#include <Storages/DeltaMerge/Segment.h>
+#include <Storages/DeltaMerge/VersionChain/RowKeyFilter.h>
+
+#include <ranges>
+
+namespace DB::DM
+{
+namespace
+{
+template <ExtraHandleType HandleType>
+UInt32 buildRowKeyFilterVector(
+    const ColumnView<HandleType> & handles,
+    const RowKeyRanges & delete_ranges,
+    const RowKeyRanges & read_ranges,
+    const UInt32 start_row_id,
+    BitmapFilter & filter)
+{
+    UInt32 filtered_out_rows = 0;
+    for (UInt32 i = 0; i < handles.size(); ++i)
+    {
+        auto in_range = [h = handles[i]](const RowKeyRange & range) {
+            return inRowKeyRange(range, h);
+        };
+        // IN delete_ranges or NOT IN read_ranges
+        if (std::any_of(delete_ranges.begin(), delete_ranges.end(), in_range)
+            || std::none_of(read_ranges.begin(), read_ranges.end(), in_range))
+        {
+            filter[i + start_row_id] = 0;
+            ++filtered_out_rows;
+        }
+    }
+    return filtered_out_rows;
+}
```

Comment:
This seems to be a critical path. Is `handles` sorted? If it is, maybe we can do some optimization.

Reply:
In DMFile, `handles` is sorted? And there are min-max handle in DMFile.